### ISSUE TITLE
Picon issue on 0.8W & 1W

### ISSUE THREAD

**Issue by decaturguy** [FernetMenta/vdr-plugin-vnsiserver#26]
Archlinux 64-bit, VDR 2.2.0 & vnsiserver 1.2.1
Test channels:
SVT1 HD;Telenor:10903:VC34M5O25S1:S1.0W:25000:1034=27:3024=sve@3;4144=sve@106:6204:B00:3801:70:69:0
NOVA Cinema;UPC Direct:11727:VC78M2S0:S0.8W:28000:151=2:161=cze@4:170:1815,D02,D97,653:30102:1536:701:0
Hi!
I'm having trouble to get picon to work with this 2 satellite, i don't know if it's bug or new future. I haven't been playing with vdr for awhile.
This used to work, generated using Pipelka's serviceref tool.
SVT1 HD;Telenor - 1_0_19_ED9_45_46_7120000_0_0_0
NOVA Cinema;UPC Direct - 1_0_1_7596_2BD_600_7100000_0_0_0

Correctly generated enigma hashes: (can you make work with this?) .
1_0_19_ED9_45_46_E060000_0_0_0    --------  SVT1 HD
1_0_1_7596_2BD_600_E080000_0_0_0  --------  NOVA Cinema

But i managed to get it work with 0.8W with this hash FFF80000
NOVA Cinema: 1_0_1_7596_2BD_600_FF80000_0_0_0
On 1W i tried many hashes like FFFA0000 nothing works.
I got the idea from this tvheadend bug report, they had almost same problem which seems
to be fixed. This links contain great info like how they fixed it and links to enigma source code.
https://tvheadend.org/issues/2700

This script generates enigma compatible hash:
https://github.com/picons/picons-source

**Comment by FernetMenta** [comment]
so what is correct now? one the one hand you say that the enigma strings are corrent:
1_0_1_7596_2BD_600_E080000_0_0_0 -------- NOVA Cinema

on the other hand you say that you made it work with:
NOVA Cinema: 1_0_1_7596_2BD_600_FF80000_0_0_0

**Comment by decaturguy** [comment]
It's suppose to be 1_0_1_7596_2BD_600_E080000_0_0_0 just like it is on a enigma box.

By reading this the first post: https://tvheadend.org/issues/2700
The guy said tvheadend generated FFF70000, so i thought maybe if i switched the 7 in the hash code to 8 i can get it to work. So i renamed my picons and it worked with FFF80000 on 0.8W, it was just a lucky guess. I have no idea what made work and how it calculated to this hash. 

So basically the plugin is not accepting enigma generated hash on 1W and on 0.8W.
Please look at your math for generating hash, Sadly i can't help you there since i can't program and i suck at math but i think that links provide the math needed for this problem.

If you provide a patch i would be happy to test it for you.

**Comment by FernetMenta** [comment]
can you test this https://github.com/FernetMenta/vdr-plugin-vnsiserver/pull/28

**Comment by decaturguy** [comment]
It didn't work :(
0.8W: E080000 failed
1.0W: E060000 failed
Just in case i tested hash generated with xvdr serviceref tool it failed too. Weired but FFF80000 still works on 0.8W.

I don't know if it just me i hope someone else can test these.
I'll post the patch that i used against 1.2.1, to make sure that i did it right.

--- vdr-plugin-vnsiserver-1.2.1/vnsiclient.c_old	2015-01-25 12:36:21.000000000 +0100
+++ vdr-plugin-vnsiserver-1.2.1/vnsiclient.c	2015-04-13 20:25:15.852273107 +0200
@@ -2303,7 +2303,6 @@
   return true;
 }
 
-// this method is taken from XVDR
 cString cVNSIClient::CreatePiconRef(cChannel* channel)
 {
   int hash = 0;
@@ -2313,16 +2312,11 @@
     hash = channel->Source() & cSource::st_Pos;
 
 #if VDRVERSNUM >= 20101
-    hash = -hash;
-#endif
-
-    if(hash > 0x00007FFF)
-      hash |= 0xFFFF0000;
-
     if(hash < 0)
-      hash = -hash;
-    else
-      hash = 1800 + hash;
+    {
+      hash = 3600 - hash;
+    }
+#endif
 
     hash = hash << 16;
   }

**Comment by FernetMenta** [comment]
did you build the picons as described here? https://github.com/picons/picons-source
you should not use any other scripts

**Comment by decaturguy** [comment]
Yes, i built using that tool but kodi is not displaying picons on 0.8W & 1W.
The tool generated and created picons as it should.
./1-build-servicelist.sh
created servicelist_vdr file:
1_0_1_79E3_2C8_600_E080000_0_0_0  tv/discoveryscience  Discovery Science
1_0_1_79E2_2C8_600_E080000_0_0_0  tv/discoveryworld    Discovery World
1_0_1_A8E_2C_46_E060000_0_0_0     tv/svt1              SVT1

Content of my srindex file:
79E3_2C8_600_E080000=tv/discoveryscience
79E2_2C8_600_E080000=tv/discoveryworld
A8E_2C_46_E060000=tv/svt1

./2-build-picons.sh
created the picons, and i extracted to ~/Bilder/vnsi (same dir is set in vnsi plugin settings in kodi)

**Comment by FernetMenta** [comment]
I have push another fix. can you rebase to https://github.com/FernetMenta/vdr-plugin-vnsiserver/tree/picons and try again.

**Comment by decaturguy** [comment]
Unfortunately it didn't work.

**Comment by FernetMenta** [comment]
please try again. now I have tested it and it works here

**Comment by decaturguy** [comment]
It doesn't work for me, did you tested with channel list with 0.8 & 1W position?

**Comment by FernetMenta** [comment]
hmm, I tested the 0.8W channel (was the only one in my list) and got E08000

**Comment by FernetMenta** [comment]
ahh, there is one digit missing ...

**Comment by FernetMenta** [comment]
pushed next fix in this try and error game :)

**Comment by decaturguy** [comment]
Almost there... :) 
It didn't work with script generated hash BUT looking at you comment you wrote E08000 missing a 0 at the end so i tried that and it worked (E08 & E06 with 3 zeros) but it suppose to have 4 zeros right?

**Comment by decaturguy** [comment]
It's working, sorry i just double checked since I'm patching against 1.2.1 i used old patch file which included "hash = hash << 20" instead of 16.

You can close this issue.

Thank you so much :)

**Comment by FernetMenta** [comment]
thanks, I merged the fix
